google/gax/__init__.py

```diff
@@ -482,3 +489,160 @@ def __next__(self):
         if self._index >= len(self._current):
             self._current = None
         return resource
+
+
+def _from_any(pb_type, any_pb):
+    if not any_pb.Is(pb_type.DESCRIPTOR):
+        raise TypeError(
+            'Could not convert {} to {}'.format(
+                any_pb.__class__.__name__, pb_type.__name__))
+
+    return pb_type.FromString(any_pb.value)
+
+
+class ResultError(GaxError):
+    """Thrown when an operation completes with an error result.
+    """
+
+    @classmethod
+    def from_operation_error(cls, operation_error):
+        """Factory: construct instance from an operation error result.
+
+        Args:
+            operation_error (google.rpc.Status): the error result of the
+                operation.
+        """
+        return cls(operation_error.message, operation_error.code)
+
+    def __init__(self, msg, code=None):
+        """Constructor.
+
+        Args:
+            msg (string): describes the error that occurred.
+            code (int, optional): the status code, which should be an enum
+                value of google.rpc.Code.
+        """
+        super(ResultError, self).__init__(msg)
+        self.code = code
+
+    def __str__(self):
+        msg = super(ResultError, self).__str__()
+        return 'ResultError({}, with code {})'.format(msg, self.code)
+
+
+class OperationFuture(object):
+    """A Future which polls a service for completion via OperationsApi."""
+
+    def __init__(self, operation, client, result_type, metadata_type,
+                 call_options=None):
+        """Constructor.
+
+        Args:
+            operation (google.longrunning.Operation): the initial long-running
+                operation.
+            client (google.cloud.gapic.longrunning.operations_api.OperationsApi):
+                the client used to manage the long-running operation with an API
+                service.
+            result_type (type): the class type to be unpacked from the result.
+            metadata_type (type): the class type to be unpacked from the
+                metadata.
+            call_options (google.gax.CallOptions, optional): the call options
+                that are used when reloading the operation.
+        """
+        self._first_operation = operation
+        self._last_operation = operation
+        self._client = client
+        self._result_type = result_type
+        self._metadata_type = metadata_type
+        self._call_options = call_options
+        self._cancelled = False
+
+    def cancel(self):
+        """If last Operation's value of `done` is true, returns false;
+           otherwise, issues OperationsApi.cancel_operation and returns true.
+        """
+        if not self._last_operation.done:
+            self._client.cancel_operation(self._last_operation.name)
+            self._cancelled = True
+
+        return self._cancelled
```

Comment:
@lukesneeringer I think we're all in agreement. I'm completely okay with the implementation that @eoogbe has for this method (`cancel()`); it matches the `c.f.Future` doc and it's non-blocking.

I'm just not sure it's worth cacheing `self._cancelled` here, because the `cancelled()` method needs to check the Operations API to see if the cancellation actually happened. If the `cancelled()` method finds out that it did, then it can cache a `self._cancelled` value.